[JYL27] iP

JYL27 opened the pull request:
## Colress
> "It looks like you're ready, then! OK! Let us begin!" - Colress ([source](https://bulbapedia.bulbagarden.net/wiki/Colress/Quotes#Pok%C3%A9mon_Black_2_and_White_2))

Colress frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- *super* fast to use
- **FREE!!!**

All you need to do is,
1. download it :envelope_with_arrow:
2. double-click it :computer_mouse:
3. add your tasks :writing_hand:
4. let it manage your tasks for you! :card_index_dividers:

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events 
- [ ] Reminders (coming soon)

Here's a sneak peek of the program code😉:
```
public static void main(String[] args) {
    new Colress("task.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JYL27 committed:
Replaced Duke.java with Colress.java. Updated greeting and exit messages.

JYL27 committed:
Added echo method and class-level fields to initialise scanner, receive and store inputs, initialise messages and detect exit command.

JYL27 committed:
Added a class-level ArrayList to store input strings. Added listEmptyMessage field to display a unique message if there are no inputs stored yet. Added a listCommand field to detect list command. Replaced echo method with makeList method and added printList method which iterates over the ArrayList and prints the entries in list format.

JYL27 committed:
Added new Task class to encapsulate Task object. Each Task has a description field of type String and isDone field of type boolean. Changed ArrayList generic typing from String to Task. Added getInput method that obtains input from user and assigns it to a class-level input field. Added addTask and editTask methods, and more command fields to facilitate adding items to list and checking and unchecking items in the list. Added invalidCommandMessage field to detect any non-command inputs.

JYL27 committed:
Renamed toDoList to taskList. Changed makeList method and addTask method to accomodate for all three types of tasks.

JYL27 committed:
Updated runtest.bat to input the contents in input.txt into Colress.java. Updated EXPECTED.TXT file and input.txt file for testing. Updated spacer field in Colress.java to be longer so that all outputs are contained within the spacers. Fixed a bug in printList method in Colress.java where an unintended output is obtained when attempting to print an empty list.

JYL27 committed:
Added two new checked exception classes to catch unknown commands and unknown task types. Added try-catch blocks in Colress.java to handle exceptions. Tweaked messages to make Colress sound less like a servant.

JYL27 committed:
Added deleteCommand to facilitate deleting items. Updated editList method to check for deletion cases.

JYL27 committed:
Updated UnknownCommandException.java to include delete in the command list. Updated test cases and expected results to test for deletion of items. Added full stops in Colress' messages

JYL27 committed:
Added class-level String fields to instantiate prompts and messages. Added more test cases.

JYL27 committed:
Update code such that it adheres to the Java Coding Standard.

Constants have all been changed to uppercase with underscores to separate words. Constants that are related also share a prefix instead of a suffix.
Constants have also been rearranged such that all related constants are grouped together.
taskList ArrayList object has been changed to tasks.
Switch statements have been changed such that there is no indentation for the case statements.

JYL27 committed:
Update EXPECTED.TXT and update getInput method to convert all input to lower case.

JYL27 committed:
Add task saving functionality to Colress.

Colress now writes tasks into a task.txt file, allowing users to save their task list when they exit the program.
The next time the program is run, Colress reads the task.txt file and repopulates the tasks ArrayList, or creates a new task.txt file if an existing task.txt file does not exist.

Update toString values of each task class such that delimiters are used.

This allows Colress to read the task.txt file and repopulate the tasks ArrayList

Update input.txt and EXPECTED.TXT to reflect the changes made to the task classes

JYL27 committed:
Merge branch 'branch-Level-7'

JYL27 committed:
Update Colress such that data that is not in the expected format is handled

JYL27 committed:
Add functionality that allows Colress to recognise date and time in a more meaningful way.

Dates and times in deadline and event tasks are now stored as LocalDate and LocalTime respectively.
Each task class now has an additional toTextFile method that outputs a format similar to toString method in Level-7
toString method now outputs a format similar to before Level-7 commits

JYL27 committed:
Merge branch 'branch-Level-8'

JYL27 committed:
Update Colress and Task classes such that Colress can print out tasks that falls on a specific date, and handle invalid date or time inputs

Task and its subclasses now have the method fallsOnDate which receives a LocalDate argument and returns true if its associated date equals the LocalDate argument.

JYL27 committed:
Update Colress so that if invalid inputs are received, Colress prompts for the same data instead of exiting the current action.

JYL27 committed:
Update Colress such that when a list is empty and the date command is called, Colress tells user that the list is empty instead of first requesting for a date.

Update input.txt and EXPECTED.TXT to reflect changes in the behaviour of Colress.

Update runtest.bat so that the file can be run in the Colress folder rather than in the text-ui-test folder.

JYL27 committed:
Refactor code to incorporate more OOP elements.

Each Command now has its own class, along with new classes Ui, Parser, Storage and TaskList.
Ui handles printing messages onto the interface and controls the workflow of the program i.e. re-prompting the user if an invalid input is detected etc.
Parser handles reading user input, parsing them, and throwing exceptions or returning placeholder values if invalid inputs are received.
Storage handles loading tasks from the task.txt file into a list, and writing over the task.txt file when changes are made.
TaskList contains the list of tasks and operates on them when required.

Each Command class has a success message field and an execute method.

Remove unused constructor in Task class and its subclasses.

Update EXPECTED.TXT and input.txt to match the behaviour of the refactored program.

Fix a bug in the runtest.bat file where a file path was wrong, resulting in unexpected behaviour when running the automated tests.

JYL27 committed:
Organise classes in packages.

Colress, Ui, TaskList, Storage and Parser are in the colress package.
All Task classes are in the colress.task package.
All Command classes are in the colress.command package.
All Exception classes are in the colress.exception package.

Update runtest.bat file in order for the code to compile correctly.

JYL27 committed:
Merge branch 'branch-A-Packages'

JYL27 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

JYL27 committed:
Update build.gradle file to run the correct class file

JYL27 committed:
Update code base such that it complies with the required style.

JYL27 committed:
Update code base to ensure variable names follow convention

JYL27 committed:
Add JUnit test cases to test Storage and TaskList classes

JYL27 committed:
Update build.gradle file to change archiveBaseName

JYL27 committed:
Update code base to include JavaDoc for public methods and classes

JYL27 committed:
Merge branch 'branch-A-JavaDoc'

JYL27 committed:
Update code base such that they follow the standard convention

JYL27 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/colress/TaskList.java
#	src/main/java/colress/Ui.java
#	src/main/java/colress/task/Deadline.java
#	src/main/java/colress/task/Event.java

JYL27 committed:
Update code base to add find functionality to Colress

JYL27 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/colress/Colress.java
#	src/main/java/colress/Parser.java
#	src/main/java/colress/Ui.java
#	src/main/java/colress/task/Task.java

JYL27 committed:
Update test files to test find functionality

jason-chuah commented on src/main/java/colress/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+ * Represents the TaskList of the Colress chatbot.
+ */
+public final class TaskList {
+    private final ArrayList<Task> TASKS;
```
Is there a reason TASKS is in all caps?

JYL27 replied:
Changed to lowercase. I wrongly characterised the variables as constants .Thanks for the feedback👍.

jason-chuah commented on src/main/java/colress/Ui.java:
```diff
@@ -0,0 +1,275 @@
+     * the user.
+     * The method checks against the TaskList object if the input is valid, and alerts the user if it is not.
+     */
+    public int promptTaskNumber(TaskList taskList) {
```
I like the way you've implemented all the different prompts into your ui

jason-chuah commented on src/main/java/colress/exception/UnknownCommandException.java:
```diff
@@ -0,0 +1,21 @@
+                list: print out your list of tasks.
+                date: print out deadlines and events that occur on a specified date.
+                find: print out tasks that contain a specified keyword.
+                bye: bid me farewell."""
```
I like the command guide here as well

jason-chuah commented on src/main/java/colress/task/Deadline.java:
```diff
@@ -0,0 +1,55 @@
+ * whether the task is marked as done or not and a LocalDate field representing its deadline.
+ */
+public class Deadline extends Task {
+    private final LocalDate DEADLINE;
```
I noticed you have a lot of your attribute names in all caps, but to my understanding attributes like deadline don't qualify as constants

JYL27 replied:
Changed to lowercase. I wrongly characterised the variables as constants .Thanks for the feedback👍.

jason-chuah reviewed (commented):
Your code seems really clean (much cleaner than mine at least), the only thing I'm not sure about is your choice of having certain variables in all caps.

JYL27 committed:
Update code base to adhere to coding standard

JYL27 committed:
Merge branch 'branch-A-CheckStyle'

JYL27 committed:
Update code base to add Enums to represent task type

JYL27 committed:
Update Storage file to make use of task type Enum

JYL27 committed:
Update Storage class such that it makes use of task type Enum

JYL27 committed:
Update code base to implement JavaFX

JYL27 committed:
Merge branch 'branch-Level-10'

JYL27 committed:
Update JavaDocs for the classes and methods

JYL27 committed:
Add css files and edited GUI displays

JYL27 committed:
Fix a bug

JYL27 committed:
Add ability to enter multiple numbers for commands using varargs

JYL27 committed:
Merge branch 'branch-A-Varargs'

NicholasCWK commented on src/main/java/colress/command/AddCommand.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Represents the add command that add a task to the list of tasks.
+ */
+public final class AddCommand extends Command {
```
Not sure if this is valid but at first glance, i thought that "AddCommand" sounds like a method name to add a command because it starts with a verb 😅

NicholasCWK commented on src/main/java/colress/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
+    public Task get(int index) {
```
I feel that this method name is a little vague, perhaps you could call it getTask?

NicholasCWK commented on src/main/java/colress/TaskList.java:
```diff
@@ -0,0 +1,155 @@
+    public Task get(int index) {
+        return tasks.get(index);
+    }
+    public int size() {
```
According to the naming conventions, method names should be verbs, perhaps you could call it getSize?

NicholasCWK commented on src/main/java/colress/command/DeleteCommand.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Represents the delete command that removes a task from the list of tasks.
+ */
+public final class DeleteCommand extends Command {
```
Same thing with the AddCommand, this kind of sounds like a method to delete a command 😅

NicholasCWK reviewed (commented):
Overall I found your code to be clear and well structured. I found your code standard very consistent, one thing to take note is that the names of some classes sound like names for methods, but these might just be nitpicks.

Lastly, good job on the detailed JavaDoc and the implementation of the various exceptions!

JYL27 committed:
Update two method names in TaskList class

JYL27 committed:
Add assertions to TaskList and Ui classes

In the Ui class, an assertion is made in the processDescription method
that the current command is an add command, since the only command
that will lead to the method being called is the AddCommand command.

In the TaskList class, assertions are made in some of the methods, that
indexes provided are not out of bounds, since whether the index is
out of bounds is already checked by the Ui class.

JYL27 committed:
Update code base to improve code quality

repopulateTask method in the Storage class has been refactored into
two more separate methods to make each method more readable

processCancel method is removed since it is not in use.

checkTask and uncheckTask methods share almost the same
lines of code.

A new method processTask is therefore created to abstract out the
logic of both methods to reduce code duplication.

Similarly, the overloaded retrieveTasks methods share very almost
the same lines of code.

A new method buildListofTasks is therefore created to abstract out the
logic of all three methods to reduce code duplication.

JYL27 committed:
Merge pull request #1 from JYL27/branch-A-Assertions

Add assertions to TaskList and Ui classes

JYL27 committed:
Merge branch 'master' of https://github.com/JYL27/ip

* 'master' of https://github.com/JYL27/ip:
  Add assertions to TaskList and Ui classes

JYL27 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/colress/TaskList.java

JYL27 committed:
Merge pull request #2 from JYL27/branch-A-CodeQuality

Update code base to improve code quality

JYL27 committed:
Merge branch 'master' of https://github.com/JYL27/ip

JYL27 committed:
Update Storage class to utilise streams

Currently, Storage task uses a for loop to write tasks
to the task file.

Using streams instead makes the code cleaner and shorter.

JYL27 committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/colress/TaskList.java

JYL27 committed:
Merge pull request #3 from JYL27/branch-A-Streams

Update Storage class to utilise streams

JYL27 committed:
Merge branch 'master' of https://github.com/JYL27/ip

JYL27 committed:
Fix a bug involving the delete command

Currently, delete command deletes tasks in order of task
number provided by the user.

However, deleting a task will change indexes of other tasks.
Therefore, some tasks end up not getting, or the wrong task
will be deleted.

Therefore, by sorting the array of task numbers, we can then
delete in descending order of task numbers so that no task
that should be deleted have their indexes changed.

Remove unused methods in TaskList class

JYL27 committed:
Create github-actions-demo.yml

JYL27 committed:
Update github-actions-demo.yml

JYL27 committed:
Fix issues with CI

JYL27 committed:
Fix issues with CI

JYL27 committed:
Fix issues with CI

JYL27 committed:
FIx issue with CI

JYL27 committed:
Fix issues with CI

JYL27 committed:
Add missing Javadocs to classes and methods

JYL27 committed:
Add missing Javadocs to Colress class

JYL27 committed:
Remove test text files from gitignore

JYL27 committed:
FIx issue with jar file

JYL27 committed:
Add more personality to Colress

Response strings and error messages have been edited to more closely mimic his dialogs in the Pokemon games

getDukeDialog method in DialogBox changed to getColressDialog instead to reflect the name of the chatbot

JYL27 committed:
Merge branch 'branch-A-Personality'

JYL27 committed:
Add error handling for end time input

Currently, end time input process does not check that end time
of an event is chronologically after its start time.

This leads to possibly illogical time periods for events.

Adding error handling for end time input prevents
such illogical time periods for events.

JYL27 committed:
Merge branch 'branch-A-MoreErrorHandling'

JYL27 committed:
Update test cases

JYL27 committed:
Update Gui

Currently, all responses by Colress have the same
background.

This makes the interface rather bland and the commands
carried out are not fleshed out as much.

By adding different coloured backgrounds that correspond
to certain command types, it makes the interface more
pleasing to look at as a whole, and the colour coding
makes the commands and error messages stand out
more.

JYL27 committed:
Merge branch 'branch-A-BetterGui'

JYL27 committed:
Edit JavaDocs to adhere to coding standard

JYL27 committed:
Add User Guide for Colress

JYL27 committed:
Merge branch 'branch-A-UserGuide'

JYL27 committed:
Edit JavaDocs

JYL27 committed:
Make Find Command case-insensitive

JYL27 committed:
Handle corrupted files better

Currently, if a file is corrupted, Colress prompts user
to delete the file, rather than creating a new file.

This might not be user-friendly for text-based users,
therefore Colress now creates a new task file if a
corrupted file is detected.

JYL27 committed:
Edit error message assertequals tests

JYL27 committed:
Edit error message in assertequal tests

JYL27 committed:
Fix bug in writing to task file

JYL27 committed:
Add Beginner and Advanced modes for command input

JYL27 committed:
Add ability to toggle between Beginner and Advanced modes

JYL27 committed:
Add new Javadocs and update existing Javadocs

JYL27 committed:
Add advanced command input functionality

JYL27 committed:
Update splicing of input

JYL27 committed:
Fix Storage bug

JYL27 committed:
Update Invalid Command format message

JYL27 committed:
Fix Add Command bug

JYL27 committed:
Fix Colress not displaying error message for invalid command format

JYL27 committed:
Merge branch 'branch-Modes'

JYL27 committed:
Add test cases for Task Classes

JYL27 committed:
Add test cases for command classes

JYL27 committed:
Add test cases for Parser class

JYL27 committed:
Reorganise test cases in ParserTest.java

JYL27 committed:
Refine code quality

JYL27 committed:
Merge branch 'master' into branch-Tests

JYL27 committed:
Package TestUtil and test text files

JYL27 committed:
Refactor and add test utility constants

JYL27 committed:
Update test cases for TaskList class

JYL27 committed:
Update test cases for Storage Class

JYL27 committed:
Fix ParserTest bug

JYL27 committed:
Update equals method in Task Classes

JYL27 committed:
Add test cases for TaskList Class

JYL27 committed:
Add Javadoc for stub class constructor

JYL27 committed:
Fix checkstyle violations

JYL27 committed:
Add Parsable interface

JYL27 committed:
Declare dependency for shadowJar

JYL27 committed:
Package Storage and TaskList classes

JYL27 committed:
Add Storable Interface

JYL27 committed:
Add TaskListable Interface

JYL27 committed:
Merge branch-Tests with master

JYL27 committed:
Clean up merge commit

JYL27 committed:
Add test cases for Colress class

JYL27 committed:
Refactor command word test constants

JYL27 committed:
Add test cases for Colress class

JYL27 committed:
Remove unused import from ColressTest.java

JYL27 committed:
Refactor execution messages for commands

JYL27 committed:
Fix checkstyle violation in ParserTest.java

JYL27 committed:
Add test cases for ExitCommand

JYL27 committed:
Add test cases for ToggleCommand

JYL27 committed:
Refactor Command classes

JYL27 committed:
Refactor UiAdvanced and UiBeginner

JYL27 committed:
Add abstract classes UiBeginner and UiAdvanced

JYL27 committed:
Add Chatbot abstract class

JYL27 committed:
Refactor Colress usages

JYL27 committed:
Add stub classes for testing

JYL27 committed:
Update test cases to use stub classes

JYL27 committed:
Remove unused fields in ToggleCommandTest

JYL27 committed:
Remove unused imports in ToggleCommandTest

JYL27 committed:
Remove non-stub class in ExitCommandTest

JYL27 committed:
Removed unused getter in Chatbot

JYL27 committed:
Update constructor in ColressStub

JYL27 committed:
Add exceptions to abstract method signature

JYL27 committed:
Add test cases for CheckCommand, UncheckCommand and DeleteCommand

JYL27 committed:
Remove unnecessary semicolons

JYL27 committed:
Add test cases for ListCommand, DateCommand and FindCommand

JYL27 committed:
Add test cases for invalid command format

JYL27 committed:
Remove unused import
